### ISSUE TITLE
config: feature overrides

### ISSUE THREAD

**Issue by blackboxsw** [canonical/ubuntu-advantage-client#1080]
Add the ability to toggle or turn on overrides in /etc/ubuntu-advantage/uaclient.conf.

The expectation is that each feature override would live as key-value pairs under
a top-level `features` key. 

After discussion with Rick,

This issue really just a mechanism by which we can override any machine-token directive, obligation or affordance via a machine-token overlay.

Features to call this issue complete:

- [x]  allow supplementing the machine-token.json response from contract server with other key/values presented by a new local override.json file 

- [x] to toggle the override on, provide the following in `/etc/ubuntu-advantage/uaclient.conf`
```yaml
features:
    machine-token-overlay: <path_to_json>
```

- [x] surface a UAConfig.features method to expose a dict of feature key/values configured

- [x] ua version should be augmented to report the following string if feature is set:
```
$ ua version
25.0 +machine-token-overlay
```

**Comment by lucasmoura** [comment]
@blackboxsw So it will not be possible to override a single feature right ? For example, if an user want to enable only the fips package, that will not be possible using this config. Or we should cover that as well ?

**Comment by blackboxsw** [comment]
@lucasmoura accental self-assign when I filed this.  sorry re-assigning to you.  I think the default case would be to enable operating on all --beta services. I think we want to avoid the granularity of override functionality as best we can because typical overrides should not be too flexible, easy etc. I don't want to productize this override as it feels generally like we are enabling testing, or giving early preview of functionality, not really trying to make the feature flag handling a full featured and flexible config language.

**Comment by mitechie** [comment]
To help, the experience I think we'd like to see is something like

  ua status
  ......(available things)

  ua status --all
  (all possible things)

  ua enable --beta $thing_not_available

Working towards something that can work in that way is the goal.

**Comment by lucasmoura** [comment]
@blackboxsw I just want to check with you an idea that I had regarding this issue. Right now, we have fields on the `machine-token.json` which are lists of dicts, such as:

```json

{
  "availableResources": [
    {
      "available": false,
      "name": "cc-eal"
    },
    {
      "available": true,
      "name": "esm-infra"
    },
    {
      "available": false,
      "name": "fips"
    },
    {
      "available": false,
      "name": "fips-updates"
    },
    {
      "available": true,
      "name": "livepatch"
    },
    {
      "available": true,
      "name": "support"
    }
  ],
}
```
Now, suppose we want to overlay that by stating that `livepatch`is not available. Our json overlay file should look like this:

```json

{
  "availableResources": [
    {
      "available": false,
      "name": "livepatch"
    },
  ],
}
```

The problem here is that to successfully overlay that information, I will need to run through the original `availableResources` list, find the position of the dict with name `livepatch` and them perform the modification. However, this logic will change slightly if we want to modify the `resourceEntitlements` values, which are identified by the key `type`, instead of `name`.

I can think I can code a solution that takes this into consideration, but maybe it get a little complex because of the list type values regarding some keys. For example, here is an initial idea of how my solution would look like:

```python
def depth_first_merge_dict(base_dict, other_dict):
    """Merge the contents of other dict into base_dict not only on top-level
    keys, but on all on the depths of the other_dict object. For example,
    using these values as entries for the function:

    base_dict = {"a": 1, "b": {"c": 2, "d": 3}}
    other_dict = {"b": {"c": 10}}

    Should update base_dict into:

    {"a": 1, "b": {"c": 10, "d": 3}}

    @param base_dict: The dict to be updated
    @param other_dict: The dict with information to be added into base_dict
    """
    for key, value in other_dict.items():
        if type(base_dict.get(key)) == dict and type(value) == dict:
            depth_first_merge_dict(base_dict[key], value)
        elif type(base_dict.get(key)) == list and type(value) == list:
            if key == 'availableResources':
                update_available_resources(base_dict[key], value)
            elif key == 'resourceEntitlements':
                update_resources_entitlements(base_dict[key], value)
        else:
            base_dict[key] = value
```

**Comment by blackboxsw** [comment]
@lucasmoura, I think we can avoid specific functions like update_resources_entitlements and update_available_resources functions if we just require that our overlay file somehow defines[ merge rules sort of like cloud-init does](https://cloudinit.readthedocs.io/en/latest/topics/merging.html). 

I expect you and I will be the primary users of the overlay during to test specific aspects of a single service at a given time. So, I think it might be easier for us to hang a "merge" key somewhere in the scope of the overlay which will tell the depth_first_merge how to merge data types we provide either as "extend" or "replace".  The default merge behavior for dict and list types would be to "extend".

Here's an example of what I was thinking:

```
# to replace "b" fully as well as extend top-level base dict
base_dict = {"a": 1, "b": [{"b1": "bv1"}, {"b2": "bv2"}], "c": ["c1", "c2"]}
overlay_dict = {"merge": {"list": "replace"}, "b": [{"b3": "bv3"}], "d": {"d1": "dv1"}}

result =  {"a": 1, "b": [{"b3": "bv3"}], "c": ["c1", "c2"], "d": {"d1": "dv1"}}

merge_depth_first(base_dict, overlay_dict):
   merge_rules = overlay_dict.get("merge", {})
   ....
   for key, value in other_dict.items():
       if key == "merge":
            continue
       if isinstance(base_dict.get(key), dict):
          if merge_rules.get("dict", "extend") == "extend":
               depth_first_merge_dict(base_dict[key], value)
       elif isinstance(base_dict.get(key), list):
          if merge_rules.get("list", "extend") == "extend":
               depth_first_merge_dict(base_dict[key], value)
        # Assume replace for all fall through cases
        base_dict[key] = value
```


What do you think about this approach? Do you think that would give us the ability to provide specific overrides at depth and avoid having to provide a full copy of the original  base_dict?

**Comment by lucasmoura** [comment]
@blackboxsw That's a really good solution, but my main concern is that it still lacks the possibility of replacing just one item of a list of dics, as we have with `availabledResources` and `resourceEntitlements`. The current solution only allows appending an item to the existing list or updating it completely. The problem is I think most of the situations would only require us to update few items of the list.

But we can assume that since we will use that for testing, that would be not problem, since we would control exactly what entilements/resources we want by just using the replace method. If we go towards that path, my concern is just that we would always need to most of the base dictionary to guarantee that the test will have all of the information that it needs. But let be a little bit clearer. For example, suppose we want to modify the `remoteServer` for livepatch:

```json
    {
          "type": "livepatch",
          "entitled": true,
          "affordances": {
            "architectures": [
              "amd64",
              "x86_64"
            ],
            "kernelFlavors": [
              "generic",
              "lowlatency",
              "oem",
              "aws"
            ],
            "minKernelVersion": "4.4",
            "series": [
              "trusty",
              "xenial",
              "bionic",
              "focal"
            ],
            "tier": "stable"
          },
          "directives": {
            "caCerts": "",
            "remoteServer": "https://livepatch.canonical.com"
          },
          "obligations": {
            "enableByDefault": true
          },
          "series": {
            "trusty": {
              "obligations": {
                "enableByDefault": false
              }
            }
          }
        }
      ]
    },
```

I would need to provided the whole dictionary again but modifying only the `remote_server` key. If I do not do that, we would put the livepatch service into a badly configured state (Although I do understand that some keys are not totally required, but some are). In that sense, it may be a little redundant.

However, I don't know if we can solve that without know which keys each list of dictionaries use as an index for finding an unique item on them (`name` for `availableResources` or `type` for `resourceEntitlements`)

**Comment by blackboxsw** [comment]
@lucasmoura I realize I'm bouncing around with potential solutions: 
I wonder if we can just represent overlay as a set of full key paths and values
What do you think about the following in an overlay JSON then to avoid redundancy:

We could encode an unique "key" lookup in the list in the key name of the overlay:

Let's work on this in a shared doc https://hackmd.io/xGdT5appRriedi_blWYJiQ
to talk through the different proposals. and come back here once we've reached an approach

**Comment by lucasmoura** [comment]
Addressed by #1104
